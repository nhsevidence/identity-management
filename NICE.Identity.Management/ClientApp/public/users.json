[
  {
    "id": "alessioisgreat",
    "firstName": "Bruce",
    "lastName": "Wayne",
    "email": "batman@goptham.com"
  },
  {
    "id": "alessioisgreat",
    "firstName": "Bruce",
    "lastName": "Wayne",
    "email": "batman@goptham.com"
  },
  {
    "id": "alessioisgreat",
    "firstName": "Bruce",
    "lastName": "Wayne",
    "email": "batman@goptham.com"
  },
  {
    "id": "alessioisgreat",
    "firstName": "Bruce",
    "lastName": "Wayne",
    "email": "batman@goptham.com"
  },
  {
    "id": "alessioisgreat",
    "firstName": "Bruce",
    "lastName": "Wayne",
    "email": "batman@goptham.com"
  },
  {
    "id": "alessioisgreat",
    "firstName": "Bruce",
    "lastName": "Wayne",
    "email": "batman@goptham.com"
  }
]
