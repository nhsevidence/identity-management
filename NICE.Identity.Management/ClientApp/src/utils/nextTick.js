/**
 *  Returns a new promise that resolves after a 0ms setTimeout, effectively clearing the call stack
 */
export const nextTick = async () => {
	return new Promise(resolve => {
		setTimeout(resolve, 0);
	});
};