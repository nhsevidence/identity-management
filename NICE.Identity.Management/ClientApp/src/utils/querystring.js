import queryString from "query-string";

export function queryStringToObject(qryString) {
	return queryString.parse(qryString);
}

export const removeQueryParameter = (url, parameter, value = null) => {
	const urlParts = url.split("?");

	if (urlParts.length < 2) return url;

	let urlParams = urlParts[1].split(/[&;]/g);

	for (let i = urlParams.length; i-- > 0; ) {
		const parParts = urlParams[i].split("=");

		if (
			parParts[0].toLowerCase() === parameter.toLowerCase() &&
			(!value ||
				(value && value.toLowerCase() === (parParts[1] || "").toLowerCase()))
		) {
			urlParams.splice(i, 1);
		}
	}

	const queryString = urlParams.length > 0 ? "?" + urlParams.join("&") : "";
	return urlParts[0] + queryString;
};

export const stripMultipleQueries = (path, queries) => {
	let strippedPath = path;

	queries.forEach((query) => {
		strippedPath = removeQueryParameter(strippedPath, query);
	});

	return strippedPath;
};

export const appendQueryParameter = (url, parameter, value) =>
	`${url}${url.indexOf("?") === -1 ? "?" : "&"}${parameter}=${value}`;
