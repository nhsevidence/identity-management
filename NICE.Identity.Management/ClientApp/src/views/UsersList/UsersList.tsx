import React, { Component } from "react";

import { Link } from "react-router-dom";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Card } from "@nice-digital/nds-card";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { PageHeader } from "@nice-digital/nds-page-header";

import { Endpoints } from "../../data/endpoints";
import { UserType } from "../../models/types";
import { Filter } from "../../components/Filter/Filter";
import { UserStatus } from "../../components/UserStatus/UserStatus";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";

type CardMetaData = {
	label?: string;
	value: React.ReactNode;
};

type UsersListProps = {};

type UsersListState = {
	data: Array<UserType>;
	error?: Error;
};

export class UsersList extends Component<UsersListProps, UsersListState> {
	constructor(props: UsersListProps) {
		super(props);
		this.state = {
			data: [],
		};
	}

	fetchData = async (url: string) => {
		let response, data;
		try {
			response = await fetch(url);
			data = await response.json();
		} catch (err) {
			let error: Error = err;

			this.setState({ error });
			return;
		}

		if (response.status === 200) {
			this.setState({ data });
		} else {
			this.setState({ error: new Error(data.message) });
		}
	};

	componentDidMount() {
		this.fetchData(Endpoints.usersList);
	}

	render() {
		const { data, error } = this.state;

		return (
			<>
				<Breadcrumbs>
					<Breadcrumb to="/" elementType={Link}>
						Home
					</Breadcrumb>
					<Breadcrumb>Users</Breadcrumb>
				</Breadcrumbs>

				<PageHeader heading="Users" />

				{!error ? (
					<Grid>
						<GridItem cols={12} md={3}>
							<Filter />
						</GridItem>
						<GridItem cols={12} md={9} aria-busy={!data.length}>
							{!data.length ? (
								<p>Loading...</p>
							) : (
								<ul className="list--unstyled">
									{data.map(user => {
										const {
											id,
											email_address,
											user_id,
											first_name,
											last_name,
										} = user;
										const usersListHeading = {
											headingText: `${first_name} ${last_name}`,
											link: {
												elementType: Link,
												destination: `/users/${id}`,
											},
										};

										const usersListMetadata: Array<CardMetaData> = [
											{
												value: <UserStatus user={user} />,
											},
											{
												label: "Email address",
												value: email_address,
											},
										];

										return (
											<Card
												{...usersListHeading}
												metadata={usersListMetadata}
												key={user_id}
											/>
										);
									})}
								</ul>
							)}
						</GridItem>
					</Grid>
				) : (
					<ErrorMessage error={error}></ErrorMessage>
				)}
			</>
		);
	}
}
