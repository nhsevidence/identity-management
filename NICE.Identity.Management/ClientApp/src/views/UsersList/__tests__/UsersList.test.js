import React from "react";
import { mount, shallow } from "enzyme";
import fetchMock from "fetch-mock";
import toJson from "enzyme-to-json";
import { UsersList } from "../UsersList";
import users from "./users.json";
import { MemoryRouter } from "react-router";
import { nextTick } from "../../../utils/nextTick";

afterEach(fetchMock.reset);

it("should show loading message before data has been loaded", () => {
	fetchMock.get("*", {});
	const wrapper = shallow(<UsersList />);
	expect(wrapper.find("p").text()).toEqual("Loading...");
});

it("should match the snapshot after data has been loaded", async () => {
	fetchMock.get("*", users);
	const wrapper = mount(
		<MemoryRouter>
			<UsersList />
		</MemoryRouter>,
	);
	await nextTick();
	wrapper.update();
	expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
});

it("should show error message when fetch returns 401 error", async () => {
	fetchMock.get("*", 401);
	const wrapper = shallow(<UsersList />);
	await nextTick();
	wrapper.update();
	expect(wrapper.find("#userslist-error")).toHaveLength(1);
});

it("should show error message when fetch returns 500 error", async () => {
	fetchMock.get("*", 500);
	const wrapper = shallow(<UsersList />);
	await nextTick();
	wrapper.update();
	expect(wrapper.find("#userslist-error")).toHaveLength(1);
});
