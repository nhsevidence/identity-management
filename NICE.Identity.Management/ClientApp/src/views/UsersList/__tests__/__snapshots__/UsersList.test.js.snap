// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot after data has been loaded 1`] = `
Array [
  Array [
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        className="breadcrumbs"
      >
        <li
          className="breadcrumbs__crumb"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            Home
          </a>
        </li>
        <li
          className="breadcrumbs__crumb"
        >
          <span>
            Users
          </span>
        </li>
      </ol>
    </nav>,
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "{
  \\"@context\\": \\"http://schema.org\\",
  \\"@type\\": \\"BreadcrumbList\\",
  \\"itemListElement\\": [
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 1,
      \\"item\\": {
        \\"@id\\": \\"/\\",
        \\"name\\": \\"Home\\"
      }
    },
    {
      \\"@type\\": \\"ListItem\\",
      \\"position\\": 2,
      \\"item\\": {
        \\"name\\": \\"Users\\"
      }
    }
  ]
}",
        }
      }
      type="application/ld+json"
    />,
  ],
  <div
    className="page-header"
    id="content-start"
  >
    <h1
      className="page-header__heading"
    >
      Users
    </h1>
  </div>,
  <div
    className="grid"
  >
    <div
      data-g="12 md:3"
    >
      <div
        className="panel"
      >
        <p>
          Filter options
        </p>
        <div
          className="input"
        >
          <label
            className="input__label"
            htmlFor="userSearch"
          >
            Search
          </label>
          <input
            className="input__input"
            id="userSearch"
            name="user-search"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      data-g="12 md:9"
    >
      <ul
        className="list--unstyled"
      >
        <article
          className="card"
        >
          <header
            className="card__header"
          >
            <p
              className="card__heading"
            >
              <a
                href="/users/usr_5457edea1b8f33391a000001"
                onClick={[Function]}
              >
                Cliff Booth
              </a>
            </p>
          </header>
          <dl
            className="card__metadata"
          >
            <div
              className="card__metadatum"
            >
              <dd>
                <span
                  className="tag tag--alpha"
                >
                  Active
                </span>
              </dd>
            </div>
            <div
              className="card__metadatum"
            >
              <dt
                className="visually-hidden"
              >
                Email address
              </dt>
              <dd>
                cliff.booth@nice.org.uk
              </dd>
            </div>
          </dl>
        </article>
        <article
          className="card"
        >
          <header
            className="card__header"
          >
            <p
              className="card__heading"
            >
              <a
                href="/users/usr_5457edea1b8f33391a000002"
                onClick={[Function]}
              >
                Vernita Green
              </a>
            </p>
          </header>
          <dl
            className="card__metadata"
          >
            <div
              className="card__metadatum"
            >
              <dd>
                <span
                  className="tag tag--alpha"
                >
                  Active
                </span>
              </dd>
            </div>
            <div
              className="card__metadatum"
            >
              <dt
                className="visually-hidden"
              >
                Email address
              </dt>
              <dd>
                vernita.green@gmail.com
              </dd>
            </div>
          </dl>
        </article>
        <article
          className="card"
        >
          <header
            className="card__header"
          >
            <p
              className="card__heading"
            >
              <a
                href="/users/usr_5457edea1b8f33391a000003"
                onClick={[Function]}
              >
                Rick Dalton
              </a>
            </p>
          </header>
          <dl
            className="card__metadata"
          >
            <div
              className="card__metadatum"
            >
              <dd>
                <span
                  className="tag tag--alpha"
                >
                  Active
                </span>
              </dd>
            </div>
            <div
              className="card__metadatum"
            >
              <dt
                className="visually-hidden"
              >
                Email address
              </dt>
              <dd>
                rick.dalton@nice.org.uk
              </dd>
            </div>
          </dl>
        </article>
        <article
          className="card"
        >
          <header
            className="card__header"
          >
            <p
              className="card__heading"
            >
              <a
                href="/users/usr_5457edea1b8f33391a000004"
                onClick={[Function]}
              >
                Vincent Vega
              </a>
            </p>
          </header>
          <dl
            className="card__metadata"
          >
            <div
              className="card__metadatum"
            >
              <dd>
                <span
                  className="tag tag--alpha"
                >
                  Active
                </span>
              </dd>
            </div>
            <div
              className="card__metadatum"
            >
              <dt
                className="visually-hidden"
              >
                Email address
              </dt>
              <dd>
                vincent.vega@nice.org.uk
              </dd>
            </div>
          </dl>
        </article>
      </ul>
    </div>
  </div>,
]
`;
